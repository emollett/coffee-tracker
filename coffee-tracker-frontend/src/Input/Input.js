import React, { Component } from 'react';
import axios from "axios";
import '../App.css';

class Input extends Component {

  // initialize our state
    state = {
      id: 0,
      message: null,
      name: null,
      coffee: null,
      datePurchased: null,
      date: null,
    };

    // our put method that uses our backend api
    // to create new query into our data base
    putDataToDB = (message, name, coffee, date, datePurchased, event) => {

      event.preventDefault(); // stop the page from refreshing

      // just a note, here, in the front end, we use the id key of our data object
      // in order to identify which we want to Update or delete.
      // for our back end, we use the object id assigned by MongoDB to modify
      // data base entries

      let currentIds = this.props.data.map(data => data.id);
      let idToBeAdded = 0;
      while (currentIds.includes(idToBeAdded)) {
        ++idToBeAdded;
      }

      axios.post("/api/putData", {
        id: idToBeAdded,
        message: message,
        name: name,
        coffee: coffee,
        date: date,
        datePurchased: datePurchased,
      });

      document.getElementById("new-coffee-form").reset();
    };

  render() {
    return (
      <form id="new-coffee-form" className="bigPadding" onSubmit={(event) => this.putDataToDB(this.state.message, this.state.name, this.state.coffee, this.state.date, this.state.datePurchased, event)}>

        <div className="smallPadding">
          <label htmlFor="coffee">Add a coffee to the database</label><br></br>
          <input
            id="coffee"
            type="text"
            required
            onChange={e => this.setState({ coffee: e.target.value })}
          />
        </div>

        <div className="smallPadding">
          <label htmlFor="name">Who bought the coffee?</label><br></br>
          <input
            id="name"
            type="text"
            autoComplete="given-name"
            required
            onChange={e => this.setState({ name: e.target.value })}
          />
        </div>

        <div className="smallPadding">
          <label htmlFor="datePurchased">What date was it purchased?</label><br></br>
          <input
            id="datePurchased"
            type="date"
            onChange={e => this.setState({ datePurchased: e.target.value })}
          />
        </div>

        <div className="smallPadding">
          <label htmlFor="dateOpened">What date was it opened?</label><br></br>
          <input
            id="dateOpened"
            type="date"
            onChange={e => this.setState({ date: e.target.value })}
          />
        </div>

        <div className="smallPadding">
          <label htmlFor="message">Add a note about the coffee</label><br></br>
          <input
            id="message"
            type="text"
            onChange={e => this.setState({ message: e.target.value })}
          />
        </div>

        <div className="smallPadding">
          <button data-testid="submitCoffeeTest" type="submit">
            <h4>Add to database</h4>
          </button>
        </div>

      </form>
    );
  }
}

export default Input;
