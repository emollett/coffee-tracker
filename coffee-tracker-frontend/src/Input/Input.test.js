import React from "react";
import ReactDOM from "react-dom";
import Input from "./Input";
import testData from "../testdata.json";
import renderer from "react-test-renderer";
import { render, fireEvent, wait, getByTestId } from "@testing-library/react";

const state = {
	id: 0,
	message: "Nice coffee",
	name: "Eleanor",
	coffee: "Tesco's finest",
	datePurchased: null,
	date: null,
};

it("renders without crashing", () => {
	const div = document.createElement("div");
	ReactDOM.render(<Input data={testData}/>, div);
	ReactDOM.unmountComponentAtNode(div);
});

it("renders correctly", () => {
	const tree = renderer
		.create(<Input data={testData}/>)
		.toJSON();
	expect(tree).toMatchSnapshot();
});

// it("submitting the form calls the function to put the data in the db", async () => {
// 	const { container } = render(<Input data={testData}/>);
// 	const coffee = container.querySelector("input[id=\"coffee\"]");
// 	const name = container.querySelector("input[id=\"name\"]");
// 	const datePurchased = container.querySelector("input[id=\"datePurchased\"]");
// 	const date = container.querySelector("input[id=\"dateOpened\"]");
// 	const message = container.querySelector("input[id=\"message\"]");
// 	const form = container.querySelector("form[id=\"new-coffee-form\"]");
// 	const putDataToDB = jest.fn();

// 	await wait(() => {
// 		fireEvent.change(coffee, {
// 			target: {
// 				value: "tesco"
// 			}
// 		});
// 	});
    
// 	await wait(() => {
// 		fireEvent.change(name, {
// 			target: {
// 				value: "Eleanor"
// 			}
// 		});
// 	});

// 	await wait(() => {
// 		fireEvent.change(datePurchased, {
// 			target: {
// 				value: null
// 			}
// 		});
// 	});
    
// 	await wait(() => {
// 		fireEvent.change(date, {
// 			target: {
// 				value: null
// 			}
// 		});
// 	});
    
// 	await wait(() => {
// 		fireEvent.change(message, {
// 			target: {
// 				value: null
// 			}
// 		});
// 	});

// 	await wait(() => {
// 		fireEvent.click(getByTestId(container, "submitCoffeeTest"));
// 	});

// 	expect(putDataToDB).toHaveBeenCalled();
// });