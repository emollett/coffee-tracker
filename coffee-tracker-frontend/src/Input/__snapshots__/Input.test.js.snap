// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<form
  className="bigPadding"
  id="new-coffee-form"
  onSubmit={[Function]}
>
  <div
    className="smallPadding"
  >
    <label
      htmlFor="coffee"
    >
      Add a coffee to the database
    </label>
    <br />
    <input
      id="coffee"
      onChange={[Function]}
      required={true}
      type="text"
    />
  </div>
  <div
    className="smallPadding"
  >
    <label
      htmlFor="name"
    >
      Who bought the coffee?
    </label>
    <br />
    <input
      autoComplete="given-name"
      id="name"
      onChange={[Function]}
      required={true}
      type="text"
    />
  </div>
  <div
    className="smallPadding"
  >
    <label
      htmlFor="datePurchased"
    >
      What date was it purchased?
    </label>
    <br />
    <input
      id="datePurchased"
      onChange={[Function]}
      type="date"
    />
  </div>
  <div
    className="smallPadding"
  >
    <label
      htmlFor="dateOpened"
    >
      What date was it opened?
    </label>
    <br />
    <input
      id="dateOpened"
      onChange={[Function]}
      type="date"
    />
  </div>
  <div
    className="smallPadding"
  >
    <label
      htmlFor="message"
    >
      Add a note about the coffee
    </label>
    <br />
    <input
      id="message"
      onChange={[Function]}
      type="text"
    />
  </div>
  <div
    className="smallPadding"
  >
    <button
      data-testid="submitCoffeeTest"
      type="submit"
    >
      <h4>
        Add to database
      </h4>
    </button>
  </div>
</form>
`;
